test/utils/tutorials/tester.go

```diff
@@ -0,0 +1,621 @@
+						if commandIndex < len(chunk.rawCommands)-1 {
+							writer.WriteString("\n")
+						}
+					}
+				}
+			}
+		}
+	}
+	return writer.Flush()
+}
+
+func findChunkById(stages [][]*ExecutableChunk, stageName string, chunkId string) *ExecutableChunk {
+	for _, stage := range stages {
+		if stage[0].Stage != stageName {
+			continue
+		}
+		for _, chunk := range stage {
+			if chunk.Id == chunkId {
+				return chunk
+			}
+		}
+	}
+	return nil
+}
+
+func (chunk *ExecutableChunk) write_bash_script(basedir string, script_name string) error {
+	scriptPath := path.Join(basedir, script_name)
+	f, err := os.Create(scriptPath)
+	defer f.Close()
+	if err != nil {
+		return err
+	}
+	writer := bufio.NewWriter(f)
+	_, err = writer.WriteString("#!/bin/bash\n")
+	if err != nil {
+		return err
+	}
+	_, err = writer.WriteString("set -euo pipefail\n")
+	if err != nil {
+		return err
+	}
+	for _, line := range chunk.trimedCommands {
+		_, err = writer.WriteString(line + "\n")
+		if err != nil {
+			return err
+		}
+	}
+	if len(chunk.Variables) > 0 {
+		for _, variable := range chunk.Variables {
+			_, err = writer.WriteString("echo " + variable + "=$" + variable + "\n")
+			if err != nil {
+				return err
+			}
+		}
+	}
+	err = writer.Flush()
+	if err != nil {
+		return err
+	}
+	return os.Chmod(scriptPath, 0770)
+}
+
+func (chunk *ExecutableChunk) createCommand(trimedCommand string, tmpDirs map[string]string, variables map[string]string) error {
+	var command RunningCommand
+	var terminatingError error
+	splited, err := shlex.Split(trimedCommand)
+	if err != nil {
+		return err
+	}
+	executable := splited[0]
+	args := splited[1:]
+	ctx := context.Background()
+	ctx, command.cancelFunc = context.WithTimeout(context.Background(), time.Duration(minutesToTimeout)*time.Minute)
+	command.cmd = exec.CommandContext(ctx, executable, args...)
+	// When the chunk requires a specific directory to be executed in
+	if chunk.RootDir == "$operator" {
+		command.cmd.Dir = operator_root
+	} else if strings.HasPrefix(chunk.RootDir, "$tmpdir") {
+		// tmpdirs are reusable between commands.
+		// so they are stored in a map.
+		dirselector := strings.Split(chunk.RootDir, string(os.PathSeparator))[0]
+		tmpdir, exists := tmpDirs[dirselector]
+		if !exists {
+			tmpdir, terminatingError = os.MkdirTemp("/tmp", "*")
+			if terminatingError != nil {
+				pterm.Error.PrintOnError(terminatingError)
+				return terminatingError
+			}
+			tmpDirs[dirselector] = tmpdir
+		}
+		command.cmd.Dir = strings.Replace(chunk.RootDir, dirselector, tmpdir, 1)
+	} else {
+		var tmpdir string
+		tmpdir, terminatingError = os.MkdirTemp("/tmp", "*")
+		// store the directory for deletion at the end of the process
+		tmpDirs[tmpdir] = tmpdir
+		if terminatingError != nil {
+			pterm.Error.PrintOnError(terminatingError)
+			return terminatingError
+		}
+		command.cmd.Dir = tmpdir
+	}
+	if chunk.Env != nil {
+		for _, v := range chunk.Env {
+			command.cmd.Env = append(command.cmd.Env, v+"="+variables[v])
```

Comment:
When the field `Env` is set the os env is skipped, see https://pkg.go.dev/os/exec. To automatically restore preserve the previous command env and avoid env/varable declarations the tutorial tester could store them in a file at the end of a command (i.e `env > cmd.env`) and restore them at the beginning of a command (i.e. `set -a && source cmd.env && set +a`)